File: extractor/extractor/strategies/kan_ifb.py
from __future__ import annotations

import re

import fitz

from extractor.models import ExtractionResult, LineItem
from extractor.strategies.base import BaseStrategy
from extractor.utils import parse_de_number

_POS_HEAD = re.compile(r"^(?P<pos>\d{3})\s+Artikelnummer:\s+(?P<art>\S+)")


class KanIfbStrategy(BaseStrategy):
    layout_id = "kan_ifb"

    def matches(self, doc: fitz.Document) -> bool:
        text = doc[0].get_text()
        return "ANGEBOT" in text and "Beleg" in text and "KAN" in text

    def extract(self, doc: fitz.Document, source_pdf: str) -> ExtractionResult:
        lines: list[str] = []
        for page in doc:
            lines.extend(line.strip() for line in page.get_text().splitlines())

        items: list[LineItem] = []
        i = 0
        while i < len(lines):
            match = _POS_HEAD.match(lines[i])
            if not match:
                i += 1
                continue
            if i + 4 >= len(lines):
                break
            qty_line = lines[i + 1]
            unit_line = lines[i + 2]
            price_line = lines[i + 3]
            total_line = lines[i + 4]
            desc_lines: list[str] = []
            j = i + 5
            while j < len(lines) and not _POS_HEAD.match(lines[j]):
                if lines[j] in ("Pos.", "Übertrag", "Betrag EUR") or lines[j].startswith(
                    "Übertrag"
                ):
                    break
                desc_lines.append(lines[j])
                j += 1
            items.append(
                LineItem(
                    position=match.group("pos"),
                    article_number=match.group("art"),
                    description=" ".join(desc_lines).strip(),
                    quantity=parse_de_number(qty_line),
                    unit=unit_line,
                    unit_price=parse_de_number(price_line),
                    line_total=parse_de_number(total_line),
                )
            )
            i = j

        return ExtractionResult(
            layout_id=self.layout_id, source_pdf=source_pdf, items=items
        )

File: desktop/src/extractor/strategies/kan_ifb.ts
import type { ExtractionResult, LineItem } from "../models";
import { parseDeNumber } from "../utils";
import { isKanAnchorLine, KAN_POS_MERGED, parseKanAnchorHead } from "../table/kan-block";

const KAN_STOP = /^(Pos\.|Übertrag|Betrag EUR|Seite\s+\d)/i;
const KAN_ALTERNATIV_MARKER = /^alternativ$/i;

export function extractFromLines(lines: string[], source_pdf: string): ExtractionResult {
  const items: LineItem[] = [];

  let i = 0;
  while (i < lines.length) {
    const line = (lines[i] ?? "").trim();
    const next = (lines[i + 1] ?? "").trim();

    if (!isKanAnchorLine(line, next)) {
      i += 1;
      continue;
    }

    const parsed = parseKanAnchorHead(
      KAN_POS_MERGED.test(line) ? [line] : [line, next],
    );
    if (!parsed) {
      i += 1;
      continue;
    }

    const dataStart = i + parsed.startIdx;
    if (dataStart + 4 > lines.length) break;

    const descLines: string[] = [];
    let j = dataStart + 4;
    while (j < lines.length) {
      const l = (lines[j] ?? "").trim();
      if (!l) {
        j += 1;
        continue;
      }
      if (KAN_STOP.test(l) || isKanAnchorLine(l, lines[j + 1])) break;
      if (KAN_ALTERNATIV_MARKER.test(l)) continue;
      descLines.push(l);
      j += 1;
    }

    items.push({
      position: parsed.position,
      article_number: parsed.article_number,
      artikel_prefix: null,
      description: descLines.join("\n").trim(),
      quantity: parseDeNumber(lines[dataStart] ?? ""),
      unit: (lines[dataStart + 1] ?? "").trim() || null,
      unit_price: parseDeNumber(lines[dataStart + 2] ?? ""),
      line_total: parseDeNumber(lines[dataStart + 3] ?? ""),
    });

    i = j;
  }

  return { layout_id: "IFB GmbH", source_pdf, items };
}
